test/common/buffer/buffer_speed_test.cc

```diff
@@ -0,0 +1,221 @@
+#include "common/buffer/buffer_impl.h"
+#include "common/common/assert.h"
+
+#include "absl/strings/string_view.h"
+#include "testing/base/public/benchmark.h"
+
+namespace Envoy {
+
+// Test the creation of an empty OwnedImpl.
+static void BufferCreateEmpty(benchmark::State& state) {
+  uint64_t length = 0;
+  for (auto _ : state) {
+    Buffer::OwnedImpl buffer;
+    length += buffer.length();
+  }
+  benchmark::DoNotOptimize(length);
+}
+BENCHMARK(BufferCreateEmpty);
+
+// Test the creation of an OwnedImpl with varying amounts of content.
+static void BufferCreate(benchmark::State& state) {
+  const std::string data(state.range(0), 'a');
+  const absl::string_view input(data);
+  uint64_t length = 0;
+  for (auto _ : state) {
+    Buffer::OwnedImpl buffer(input);
+    length += buffer.length();
+  }
+  benchmark::DoNotOptimize(length);
+}
+BENCHMARK(BufferCreate)->Arg(1)->Arg(4096)->Arg(16384)->Arg(65536);
+
+// Test the appending of varying amounts of content from a string to an OwnedImpl.
+static void BufferAddString(benchmark::State& state) {
+  const std::string data(state.range(0), 'a');
+  const absl::string_view input(data);
+  Buffer::OwnedImpl buffer(input);
+  for (auto _ : state) {
+    buffer.add(data);
+    // Drain from the front of the Buffer to keep the test's memory usage fixed.
+    // @note Ideally we could use state.PauseTiming()/ResumeTiming() to exclude
+    // the time spent in the drain operation, but those functions themselves are
+    // heavyweight enough to cloud the measurements:
+    // https://github.com/google/benchmark/issues/179
+    buffer.drain(input.size());
+  }
+  uint64_t length = buffer.length();
+  benchmark::DoNotOptimize(length);
+}
+BENCHMARK(BufferAddString)->Arg(1)->Arg(4096)->Arg(16384)->Arg(65536);
+
+// Variant of BufferAddString that appends from another Buffer::Instance
+// rather than from a string.
+static void BufferAddBuffer(benchmark::State& state) {
+  const std::string data(state.range(0), 'a');
+  const absl::string_view input(data);
+  const Buffer::OwnedImpl to_add(data);
+  Buffer::OwnedImpl buffer(input);
+  for (auto _ : state) {
+    buffer.add(to_add);
+    // Drain from the front of the Buffer to keep the test's memory usage fixed.
+    buffer.drain(input.size());
+  }
+  uint64_t length = buffer.length();
+  benchmark::DoNotOptimize(length);
+}
+BENCHMARK(BufferAddBuffer)->Arg(1)->Arg(4096)->Arg(16384)->Arg(65536);
+
+static void BufferDrain(benchmark::State& state) {
+  const std::string data(state.range(0), 'a');
+  const absl::string_view input(data);
+  const Buffer::OwnedImpl to_add(data);
+  Buffer::OwnedImpl buffer(input);
+
+  // On each iteration of the benchmark, add N bytes and drain a multiple of N, as specified
+  // by DrainCycleRatios. This exercises full-slice, partial-slice, and multi-slice code paths
+  // in the Buffer's drain implementation.
```

Comment:
My thoughts around this one is that I think it covers the behaviors you mentioned but I can imagine implementations where it might not tickle perf bugs.  

I've seen a bunch of classes where resizing got super inefficient when you cross some arbitrary threshhold (usually powers of 2 but YNK) due to high-low resizing issues.  Do you think it might be worthwhile to have an add/remove test which added/removed a large number of "random" (fixed seed for reproducability but try many values tested within a benchmark) adds and drains to make sure there aren't expensive spurious copies / allocations as we add and remove data?  

The common case for proxying is we're getting random header and body chunks from the network and parsing some subset as we pass things through the chain, so I think it'd be useful for the tests here and below.  Before you do any work I'd be interested in @jmarantz  thoughts on useful vs overkill as I think he does more perf testing than I.